### PR TITLE
denormalise name in elasticsearch; add id endpoints

### PR DESCRIPTION
**Pull request by rogpeppe**
We don't need to denormalise the id-related data in the entity mongo doc
yet, although that will be easy to add later if/when required.

### CODE REVIEW

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results: http://ci.jujugui.org:8080/job/charmstore/528/

File: internal/charmstore/search.go [hunk 1]
```diff
@@ -26,14 +26,23 @@ type StoreElasticSearch struct {
 
 const typeName = "entity"
 
+type esDoc struct {
+	*mongodoc.Entity
+	Name string
```
Comment (urosj) [inline internal/charmstore/search.go]: Please add series (as they are more important as Name) and user as well.

File: internal/charmstore/search.go [hunk 2]
```diff
@@ -26,14 +26,23 @@ type StoreElasticSearch struct {
...
 // is configured.
 func (ses *StoreElasticSearch) put(entity *mongodoc.Entity) error {
 	if ses == nil || ses.Index == nil {
 		return nil
 	}
-	_, err := ses.PutDocumentVersion(typeName, ses.getID(entity), int64(entity.URL.Revision), entity)
-	return err
+	doc := esDoc{
+		Entity: entity,
+		Name:   entity.URL.Name,
```
Comment (urosj) [inline internal/charmstore/search.go:44]: User and series as well, please.

File: internal/charmstore/search_test.go [hunk 3]
```diff
@@ -369,3 +369,10 @@ func assertSearchResults(c *gc.C, obtained SearchResult, expected []string) {
 		c.Assert(ids[i], gc.Equals, v)
 	}
 }
+
+func esDocForEntity(e *mongodoc.Entity) esDoc {
+	return esDoc{
+		Entity: e,
+		Name:   e.URL.Name,
```
Comment (urosj) [inline internal/charmstore/search_test.go:376]: User and Series.

File: internal/v4/api.go [hunk 4]
```diff
@@ -439,6 +443,30 @@ func (h *Handler) metaRevisionInfo(id *charm.Reference, path string, flags url.V
...
+func (h *Handler) metaIdName(entity *mongodoc.Entity, id *charm.Reference, path string, flags url.Values) (interface{}, error) {
+	return id.Name, nil
+}
+
+// GET id/meta/id-revision
+// http://tinyurl.com/ntd3coz
+func (h *Handler) metaIdRevision(entity *mongodoc.Entity, id *charm.Reference, path string, flags url.Values) (interface{}, error) {
+	return id.Revision, nil
+}
+
```
Comment (urosj) [inline internal/v4/api.go:477]: Ok, why not :)

**Comment by urosj** [comment]
After adding user and series to ES doc with tests, :+1:  and "ship.it" :)

File: internal/v4/api.go [hunk 5]
```diff
@@ -439,6 +443,30 @@ func (h *Handler) metaRevisionInfo(id *charm.Reference, path string, flags url.V
 	return response, nil
 }
 
+// GET id/meta/id-user
+// http://tinyurl.com/o7xmhz2
+func (h *Handler) metaIdUser(entity *mongodoc.Entity, id *charm.Reference, path string, flags url.Values) (interface{}, error) {
+	return id.User, nil
```
Comment (frankban) [inline internal/v4/api.go]: It seems we are not using the response types on those four endpoints.
Shouldn't this be something like:
return &params.User{User: id.User}, nil
?
Also note that below I'll suggest to change those type names.

File: params/params.go [hunk 6]
```diff
@@ -123,5 +123,29 @@ type SearchResponse struct {
 	Results    []SearchResult
 }
 
+// User holds the result of an id/meta/id-user GET request.
+// See http://tinyurl.com/o7xmhz2
+type User struct {
```
Comment (frankban) [inline params/params.go]: type IdUserResponse, to be consistent with other response types?

File: params/params.go [hunk 7]
```diff
@@ -123,5 +123,29 @@ type SearchResponse struct {
...
+	User string
+}
+
+// Series holds the result of an id/meta/id-series GET request.
```
Comment (frankban) [inline params/params.go]: Ditto: IdSeriesRresponse.

File: params/params.go [hunk 8]
```diff
@@ -123,5 +123,29 @@ type SearchResponse struct {
...
+// See http://tinyurl.com/pnwmr6j
+type Series struct {
+	Series string
+}
+
+// Name holds the result of an id/meta/id-name GET request.
```
Comment (frankban) [inline params/params.go]: Ditto: IdNameResponse.

File: params/params.go [hunk 9]
```diff
@@ -123,5 +123,29 @@ type SearchResponse struct {
...
+// See http://tinyurl.com/m5q8gcy
+type Name struct {
+	Name string
+}
+
+// Revision holds the result of an id/meta/id-revision GET request.
```
Comment (frankban) [inline params/params.go]: ... IdRevisionResponse.

**Comment by frankban** [comment]
:+1: with changes, thank you!

File: internal/v4/api.go (hunk 5 above)
Comment (rogpeppe) [inline internal/v4/api.go]: Good point and an excellent catch, thanks!

File: params/params.go (hunk 6 above)
Comment (rogpeppe) [inline params/params.go]: done

File: params/params.go (hunk 7 above)
Comment (rogpeppe) [inline params/params.go]: done

File: params/params.go (hunk 8 above)
Comment (rogpeppe) [inline params/params.go]: done

File: params/params.go (hunk 9 above)
Comment (rogpeppe) [inline params/params.go]: done

File: internal/charmstore/search_test.go (hunk 3 above)
Comment (rogpeppe) [inline internal/charmstore/search_test.go:376]: done.

**Comment by rogpeppe** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/charmstore-merge

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results: http://ci.jujugui.org:8080/job/charmstore/529/

**Pull request merged**